ExamplePipeline.yml

```diff
@@ -0,0 +1,132 @@
+      # Optional
+      requirements: '-r tests/requirements.txt'
+      TestReport: true
+
+  Coverage:
+    uses: pyTooling/Actions/.github/workflows/CoverageCollection.yml@dev
+    needs:
+      - Params
+    with:
+      artifact: ${{ fromJson(needs.Params.outputs.params).artifacts.coverage }}
+      # Optional
+      python_version: ${{ fromJson(needs.Params.outputs.params).python_version }}
+      requirements: '-r tests/requirements.txt'
+    secrets:
+      codacy_token: ${{ secrets.CODACY_PROJECT_TOKEN }}
+
+  StaticTypeCheck:
+    uses: pyTooling/Actions/.github/workflows/StaticTypeCheck.yml@dev
+    needs:
+      - Params
+    with:
+      mypy_args: -m ${{ fromJson(needs.Params.outputs.params).package }}
+      artifact: ${{ fromJson(needs.Params.outputs.params).artifacts.typing }}
+      # Optional
+      python_version: ${{ fromJson(needs.Params.outputs.params).python_version }}
+      requirements: '-r tests/requirements.txt'
+      html_report: 'htmlmypy'
+
+  Release:
+    uses: pyTooling/Actions/.github/workflows/Release.yml@dev
+    if: startsWith(github.ref, 'refs/tags')
+    needs:
+      - UnitTesting
+      - Coverage
+      - StaticTypeCheck
+
+  Package:
+    uses: pyTooling/Actions/.github/workflows/Package.yml@dev
+    if: startsWith(github.ref, 'refs/tags')
+    needs:
+      - Params
+      - Coverage
+    with:
+      artifact: ${{ fromJson(needs.Params.outputs.params).artifacts.wheel }}
+      # Optional
+      python_version: ${{ fromJson(needs.Params.outputs.params).python_version }}
+      requirements: 'wheel'
+
+  PublishOnPyPI:
+    uses: pyTooling/Actions/.github/workflows/PublishOnPyPI.yml@dev
+    if: startsWith(github.ref, 'refs/tags')
+    needs:
+      - Params
+      - Release
+      - Package
+    with:
+      artifact: ${{ fromJson(needs.Params.outputs.params).artifacts.wheel }}
+      # Optional
+      python_version: ${{ fromJson(needs.Params.outputs.params).python_version }}
+      requirements: 'wheel twine'
+    secrets:
+      PYPI_TOKEN: ${{ secrets.PYPI_TOKEN }}
+
+  VerifyDocs:
+    uses: pyTooling/Actions/.github/workflows/VerifyDocs.yml@dev
+    needs:
+      - Params
+    with:
+      # Optional
+      python_version: ${{ fromJson(needs.Params.outputs.params).python_version }}
+
+  BuildTheDocs:
+    uses: pyTooling/Actions/.github/workflows/BuildTheDocs.yml@dev
+    needs:
+      - Params
+      - VerifyDocs
+    with:
+      artifact: ${{ fromJson(needs.Params.outputs.params).artifacts.doc }}
+
+  PublishToGitHubPages:
+    uses: pyTooling/Actions/.github/workflows/PublishToGitHubPages.yml@dev
+    needs:
+      - Params
+      - BuildTheDocs
+      - Coverage
+      - StaticTypeCheck
+    with:
+      doc: ${{ fromJson(needs.Params.outputs.params).artifacts.doc }}
+      # Optional
+      coverage: ${{ fromJson(needs.Params.outputs.params).artifacts.coverage }}
+      typing: ${{ fromJson(needs.Params.outputs.params).artifacts.typing }}
+
+
+  ArtifactCleanUp:
```

Comment:
We need 2 steps. One for packages if not tagged and one for remaining artifacts.

Moreover, can you compute a string (`\n` separated) for the python versions based on `python_version_list` ?

```yaml
  ArtifactCleanUp:
    name: 🗑️ Artifact Cleanup
    runs-on: ubuntu-latest
    needs:
      - UnitTesting
      - Coverage
      - StaticTypeCheck
      - BuildTheDocs
      - Package
      - PublishTestResults
      - PublishToGitHubPages

    env:
#      UNITTEST: ${{ needs.UnitTesting.outputs.artifact }}-${{ needs.UnitTesting.outputs.versions }}
      COVERAGE: ${{ needs.Coverage.outputs.artifact }}
      TYPING:   ${{ needs.StaticTypeCheck.outputs.artifact }}
      DOC:      ${{ needs.BuildTheDocs.outputs.artifact }}
      PACKAGE:  ${{ needs.Package.outputs.artifact }}

    steps:
      - name: 🗑️ Delete package Artifacts
        if: ${{ ! startsWith(github.ref, 'refs/tags') }}
        uses: geekyeggo/delete-artifact@v1
        with:
          name: |
            ${{ env.PACKAGE }}
      - name: 🗑️ Delete remaining Artifacts
        uses: geekyeggo/delete-artifact@v1
        with:
          name: |
            ${{ needs.UnitTesting.outputs.artifact }}-3.6
            ${{ needs.UnitTesting.outputs.artifact }}-3.7
            ${{ needs.UnitTesting.outputs.artifact }}-3.8
            ${{ needs.UnitTesting.outputs.artifact }}-3.9
            ${{ needs.UnitTesting.outputs.artifact }}-3.10
            ${{ env.COVERAGE }}
            ${{ env.DOC }}
            ${{ env.TYPING }}
```
Source: https://github.com/pyTooling/pyTooling/blob/main/.github/workflows/Pipeline.yml#L477

Reply:
Good question ...

In the README it's not mentioned and there is no usage example with `*`.

--------------
There is https://github.com/GeekyEggo/delete-artifact/issues/1
* it was requested and discussed (04.05.2020)
* it was then implemented for testing by users (06.07.2020)
* it was confirmed to work (06.07.2020)  
  https://github.com/GeekyEggo/delete-artifact/issues/1#issuecomment-654318511)

Reply:
Now implemented.

Reply:
I updated the ArtifactCleanUp step, according to your example above. However, note that:

- I added `PublishOnPyPI` to the list of needs, because otherwise there might be a race for removing/using the package artifact.
- Should the `PublishToGitHubPages` be disabled for PRs?
- I'm not sure the ArtifactCleanUp will work as intended. If `PublishOnPyPI` or `PublishToGitHubPages` are skipped, the ArtifactCleanUp job might be skipped...

Reply:
No, the needs to `PublishOnPyPI` is wrong.

There is no race condition, because in case of a tagged run, the artifact cleanup is not executed. See the if condition for the firsts cleanup step.

In the `PublishOnPyPI` job, which runs only on tagged pipelines, there is a artifact cleanup inside to remove the artifact.

This flow is tested on pyTooling and works correct.

Reply:
If you have this needs dependency and it's not tagged, then the cleanup never runs, as the needs will not be fulfilled.

Reply:
See this YAML code:
* `PublishOnPyPI`
  https://github.com/pyTooling/pyTooling/blob/main/.github/workflows/Pipeline.yml#L270
  https://github.com/pyTooling/pyTooling/blob/main/.github/workflows/Pipeline.yml#L305-L309
* `ArtifactCleanUp`
  https://github.com/pyTooling/pyTooling/blob/main/.github/workflows/Pipeline.yml#L497-L502

Reply:
Ok. I added the deletion to job PublishOnPyPI and I removed it from the needs.